File: js-python-educative-challenges/two-pointers/reverse_words_in_a_string.js
/**
 * Given a sentence, reverse the order of its words without affecting the order
 * of letters within a given word.
 *
 * @param {string} s
 * @returns {string}
 */

function reverseWords(s) {
  rev_arr = s.replace(/ +/, ' ').trim().split('').reverse()
  let first = 0;
  let second = 0;

  while (first < rev_arr.length) {
    while (second < rev_arr.length && rev_arr[second] !== " ") {
      second++
    }
    rev_arr = [].concat(
      rev_arr.slice(0, first), rev_arr.slice(first, second).reverse(),
      rev_arr.slice(second)
    )
    first = second + 1
    second = second + 1
  }
  return rev_arr.join('')
}

// main

const s = "You   are amazing  "
console.log(reverseWords(s))

File: js-python-educative-challenges/two-pointers/reverse_words_in_a_string.py
"""Given a sentence, reverse the order of its words without affecting the order
 of letters within a given word."""

import re

def reverse_words(string: str) -> str:
    s = re.sub(' +', ' ', string[::-1].strip())
    first = 0
    second = 0

    while first < len(s):
        while second < len(s) and s[second] != ' ':
            second += 1

        s = f"{s[:first]}{s[first:second][::-1]}{s[second:]}"
        first = second + 1
        second += 1
    return s

if __name__ == '__main__':
    s = "You   are amazing  " * 10
    print(reverse_words(s))
